components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -311,23 +312,172 @@ public void deleteUser(String userId) throws NotFoundException, CharonException
                 requiredAttributes);
     }
 
-    private List<Object> listUsers(Map<String, Boolean> requiredAttributes) throws CharonException {
+    private List<Object> listUsers(Map<String, Boolean> requiredAttributes, int offset, int limit, String sortBy,
+            String sortOrder, String domainName) throws CharonException {
 
-        ClaimMapping[] coreClaims;
-        ClaimMapping[] userClaims;
-        ClaimMapping[] extensionClaims = null;
         List<Object> users = new ArrayList<>();
-        //0th index is to store total number of results;
+        //0th index is to store total number of results.
         users.add(0);
+
+        String[] userNames;
+
+        if (StringUtils.isNotEmpty(domainName)) { // When domain is not null or not empty.
+            if (offset == 1 && limit == 0) { // If pagination not enable
+                userNames = getAllUsernamesByDomainName(domainName);
+            } else {
```

Comment:
What happens if the offset or limit is less then 0, here we didn't handle it yet

Reply:
if the offset or limit is less than 0, it will handle before calling this "listUsers" method. If the offset less that or equal to 0, that will be set to 1 and if limit less than or equal to 0, it will be set to 0. Therefore we don't need to handle it inside the "listUsers" method again.